### PR TITLE
Add metrics for cross cluster implementation

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Add metrics for cross cluster components
- Add missing parts in cross-cluster components
- Handle shard level error when getting cross cluster tasks

**Why?**
- Cross-domain operations support

**How did you test it?**
- Existing test since it's mostly adding metrics

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: cmd/server/tools.go [hunk 1]
```diff
@@ -18,6 +18,7 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+//go:build tools
```
Comment (demirkayaender) [inline cmd/server/tools.go:21]: nit: is this generated?

File: common/util.go [hunk 2]
```diff
@@ -982,3 +982,17 @@ func ConvertErrToGetTaskFailedCause(err error) types.GetTaskFailedCause {
...
+func ConvertGetTaskFailedCauseToErr(failedCause types.GetTaskFailedCause) error {
+	switch failedCause {
+	case types.GetTaskFailedCauseServiceBusy:
+		return &types.ServiceBusyError{}
+	case types.GetTaskFailedCauseTimeout:
+		return context.DeadlineExceeded
+	case types.GetTaskFailedCauseShardOwnershipLost:
+		return &types.ShardOwnershipLostError{}
+	default:
+		return &types.InternalServiceError{Message: "uncategorized error"}
```
Comment (demirkayaender) [inline common/util.go:996]: Do you have the message from the failedCause so we can emit here?

File: cmd/server/tools.go (hunk 1 above)
Comment (yycptt) [reply cmd/server/tools.go:21]: Yes auto-generated by editor when running `make bins`

File: service/history/queue/cross_cluster_queue_processor_base.go [hunk 3]
```diff
@@ -276,7 +276,9 @@ processorPumpLoop:
 				continue processorPumpLoop
 			}
 
-			if c.readyForPollTasks.Len() > c.options.MaxPendingTaskSize() {
+			numReadyForPoll := c.readyForPollTasks.Len()
+			c.metricsScope.RecordTimer(metrics.CrossClusterTaskPendingTimer, time.Duration(numReadyForPoll))
+			if numReadyForPoll > c.options.MaxPendingTaskSize() {
 				c.logger.Warn("too many outstanding ready for poll tasks in cross cluster queue.")
```
Comment (demirkayaender) [inline service/history/queue/cross_cluster_queue_processor_base.go:282]: Should we make this an error log?

File: service/history/shard/context.go [hunk 4]
```diff
@@ -180,11 +180,12 @@ var (
...
 const (
-	conditionalRetryCount    = 5
-	logWarnTransferLevelDiff = 3000000 // 3 million
-	logWarnTimerLevelDiff    = time.Duration(30 * time.Minute)
-	historySizeLogThreshold  = 10 * 1024 * 1024
-	minContextTimeout        = 1 * time.Second
+	conditionalRetryCount       = 5
+	logWarnTransferLevelDiff    = 3000000 // 3 million
+	logWarnTimerLevelDiff       = time.Duration(30 * time.Minute)
+	logWarnCrossClusterLevelLag = 3000000
```
Comment (demirkayaender) [inline service/history/shard/context.go:186]: is this a duration? if so, do you mind commenting the unit of this?

File: common/util.go (hunk 2 above)
Comment (yycptt) [reply common/util.go:996]: The failed cause is an enum so we don't have the message here (the target cluster). The detailed error message will be logged at the source cluster during fetch. The target side only needs to know it's an unexpected error and retry.

File: service/history/queue/cross_cluster_queue_processor_base.go (hunk 3 above)
Comment (yycptt) [reply service/history/queue/cross_cluster_queue_processor_base.go:282]: hmmm, I don't have a strong opinion but probably not? This is expected if the target cluster is under high load and slow down the polling.

File: service/history/shard/context.go (hunk 4 above)
Comment (yycptt) [reply service/history/shard/context.go:186]: It's not duration, but an absolute number based on for taskID which is calculated based on rangeID. So it's a very rough estimate, during shard movement rangeID will increase by 1 and taskID will be increased by around 1M. I basically used the same number as `logWarnTransferLevelDiff`.

Comment (demirkayaender) [reply service/history/shard/context.go:186]: Can we use a more descriptive name like minTaskIDToLogWarn or add a comment here to explain what this is?

Comment (yycptt) [reply service/history/shard/context.go]: Added a comment for transfer and cross-cluster level

**Pull request merged**